### PR TITLE
Separate single-block cache into its own object, add a test

### PR DESCRIPTION
**Pull request by barakmich**

### CODE REVIEW

File: file_cache.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	return copy(sb.openData[from:to], data), nil
+}
+
+func (sb *singleBlockCache) sync(ctx context.Context) error {
+	if !sb.openWrote {
+		return nil
+	}
+	start := time.Now()
+	err := sb.blocks.PutBlock(ctx, sb.ref, sb.openIdx, sb.openData)
+	delta := time.Now().Sub(start)
```
Comment (xiang90) [inline file_cache.go]: delta := time.Since(start)

File: file_cache.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	return err
+}
+
+func (sb *singleBlockCache) openRead(ctx context.Context, i int) error {
+	start := time.Now()
+	d, err := sb.blocks.GetBlock(ctx, i)
+	if err != nil {
+		return err
+	}
+	delta := time.Now().Sub(start)
```
Comment (xiang90) [inline file_cache.go]: delta := time.Since(start)

File: file_cache.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	// half-finished blocks
+	openIdx   int
+	openData  []byte
+	openWrote bool
+
+	ref INodeRef
+
+	blocks Blockset
+
+	readData []byte
```
Comment (xiang90) [inline file_cache.go]:
```go
readIdx int
readData []byte
```

keep the order as same as openIdx, openData?

File: file_cache.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		sb.openData = make([]byte, sb.blkSize)
+		sb.openIdx = i
+		return nil
+	}
+	start := time.Now()
+	d, err := sb.blocks.GetBlock(ctx, i)
+	if err != nil {
+		return err
+	}
+	delta := time.Now().Sub(start)
```
Comment (xiang90) [inline file_cache.go]: delta := time.Since(start)

File: file_cache.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	if sb.openIdx == i && sb.openData != nil {
+		return nil
+	}
+	if sb.openData != nil {
+		err := sb.sync(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	if sb.blocks.Length() == i {
```
Comment (xiang90) [inline file_cache.go]: shall we panic if i > sb.blocks.Length?

File: file_cache.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+func (sb *singleBlockCache) sync(ctx context.Context) error {
+	if !sb.openWrote {
+		return nil
+	}
+	start := time.Now()
+	err := sb.blocks.PutBlock(ctx, sb.ref, sb.openIdx, sb.openData)
+	delta := time.Now().Sub(start)
+	promFileBlockWrite.Observe(float64(delta.Nanoseconds()) / 1000)
+	sb.openIdx = -1
```
Comment (xiang90) [inline file_cache.go]: why do we need to clear buf after sync?

**Comment by barakmich** [comment]
Fixed (and fixed a bug)

**Comment by xiang90** [comment]
lgtm

**Pull request merged**